File: 02/1.py
def main():
    n = 1000
    a = []
    for i in range(1000):
        a.append(list(map(int, input().split())))

    ans = 0
    for line in a:
        f = 1
        if line[1] > line[0]:
            for x, y in zip(line, line[1:]):
                if y - x > 3 or y - x < 1:
                    f = 0
        else:
            for x, y in zip(line[1:], line):
                if y - x > 3 or y - x < 1:
                    f = 0
        ans += f

    print(ans)


if __name__ == "__main__":
    main()

File: 19/1.hs
import System.IO
import Data.List (isPrefixOf)
import Data.List.Split
import Debug.Trace (trace)
import qualified Data.Map as Map

main::IO ()
main = do
  let  filename = "1.in"
  content <- readFile filename
  let ans = count content
  print  ans


parse :: String -> ([String], [String])
parse input = (tokens , liness)
  where
  [part1 , part2] = splitOn "\n\n" input
  tokens = splitOn ", " part1
  liness = lines part2

check :: String -> [String] -> Map.Map String Bool -> (Bool, Map.Map String Bool)
check "" _ cache = (True, cache)
check line tokens cache =
    case Map.lookup line cache of
        Just result -> (result, cache)
        Nothing ->
            let
                (result, newCache) = foldl checkToken (False, cache) tokens
                checkToken (found, cache') token
                    | found = (True, cache')
                    | token `isPrefixOf` line =
                        let (res, updatedCache) = check (drop (length token) line) tokens cache'
                        in (res, updatedCache)
                    | otherwise = (found, cache')
                updatedCache = Map.insert line result newCache
            in (result, updatedCache)

checkWithMemo :: String -> [String] -> Bool
checkWithMemo line tokens = fst $ check line tokens Map.empty

count :: String -> Int
count input = length $ filter (`checkWithMemo` tokens) liness
  where
    (tokens, liness) = parse input
